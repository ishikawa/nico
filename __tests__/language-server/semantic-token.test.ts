import {
  LanguageServer,
  NotificationMessage,
  buildInitialize,
  buildInitialized,
  buildTextDocumentDidOpen,
  buildTextDocumentSemanticTokenFull
} from "../util/lsp";

let server: LanguageServer | undefined;

beforeAll(async () => {
  server = LanguageServer.spawn();

  // initialize
  {
    const request = buildInitialize();
    await server.sendRequest(request);
  }

  // initialized
  {
    const notification = buildInitialized();
    await server.sendNotification(notification);
  }
});

afterAll(() => {
  if (server) {
    return server.stop();
  }
});

test("open a document", async done => {
  const uri = "file:///home/user/nico/sample.nico";

  // Open document and no compilation errors
  {
    const nextNotification = server!.nextMessage<NotificationMessage>();
    const notification1 = buildTextDocumentDidOpen(uri, "1 + 2");
    await server!.sendNotification(notification1);

    const notification2 = await nextNotification;
    expect(notification2).toEqual({
      jsonrpc: "2.0",
      method: "textDocument/publishDiagnostics",
      // no errors
      params: { diagnostics: [], uri }
    });
  }

  // Semantic coloring
  {
    const request = buildTextDocumentSemanticTokenFull(uri);
    const response = await server?.sendRequest(request);

    expect(response).toEqual({
      id: request.id,
      jsonrpc: "2.0",
      result: {
        data: [0, 0, 1, 3, 0, 0, 2, 1, 4, 0, 0, 2, 1, 3, 0]
      },
      error: null
    });
  }

  done();
});
