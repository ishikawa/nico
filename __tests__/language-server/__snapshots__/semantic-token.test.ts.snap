// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0: open a document at \`-\` 1`] = `
Object {
  "error": null,
  "id": 1001,
  "jsonrpc": "2.0",
  "result": Object {
    "data": Array [],
  },
}
`;

exports[`1: open a document at \`1\` 1`] = `
Object {
  "error": null,
  "id": 1001,
  "jsonrpc": "2.0",
  "result": Object {
    "data": Array [
      0,
      0,
      1,
      3,
      0,
    ],
  },
}
`;

exports[`2: open a document at \`1 + 2\` 1`] = `
Object {
  "error": null,
  "id": 1001,
  "jsonrpc": "2.0",
  "result": Object {
    "data": Array [
      0,
      0,
      1,
      3,
      0,
      0,
      2,
      1,
      4,
      0,
      0,
      2,
      1,
      3,
      0,
    ],
  },
}
`;

exports[`3: open a document at \`./samples/fib.nico\` 1`] = `
Object {
  "error": null,
  "id": 1001,
  "jsonrpc": "2.0",
  "result": Object {
    "data": Array [
      0,
      0,
      48,
      5,
      0,
      1,
      0,
      6,
      0,
      0,
      0,
      7,
      3,
      0,
      0,
      0,
      4,
      3,
      8,
      0,
      0,
      4,
      1,
      9,
      0,
      1,
      4,
      2,
      0,
      0,
      0,
      3,
      1,
      9,
      0,
      0,
      2,
      2,
      4,
      0,
      0,
      3,
      1,
      3,
      0,
      1,
      8,
      1,
      9,
      0,
      1,
      4,
      4,
      0,
      0,
      1,
      8,
      3,
      8,
      0,
      0,
      4,
      1,
      9,
      0,
      0,
      2,
      1,
      4,
      0,
      0,
      2,
      1,
      3,
      0,
      0,
      3,
      1,
      4,
      0,
      0,
      2,
      3,
      8,
      0,
      0,
      4,
      1,
      9,
      0,
      0,
      2,
      1,
      4,
      0,
      0,
      2,
      1,
      3,
      0,
      1,
      4,
      3,
      0,
      0,
      1,
      0,
      3,
      0,
      0,
      2,
      0,
      3,
      0,
      0,
      0,
      4,
      10,
      8,
      0,
      0,
      11,
      1,
      9,
      0,
      0,
      3,
      1,
      9,
      0,
      1,
      4,
      11,
      1,
      0,
      0,
      12,
      3,
      8,
      0,
      0,
      4,
      1,
      9,
      0,
      1,
      4,
      2,
      0,
      0,
      0,
      3,
      1,
      9,
      0,
      0,
      4,
      1,
      9,
      0,
      1,
      8,
      10,
      8,
      0,
      0,
      11,
      1,
      9,
      0,
      0,
      2,
      1,
      4,
      0,
      0,
      2,
      1,
      3,
      0,
      0,
      3,
      1,
      9,
      0,
      1,
      4,
      3,
      0,
      0,
      1,
      0,
      3,
      0,
      0,
      2,
      0,
      3,
      0,
      0,
      0,
      4,
      9,
      8,
      0,
      0,
      10,
      1,
      9,
      0,
      1,
      4,
      10,
      8,
      0,
      0,
      11,
      1,
      3,
      0,
      0,
      3,
      1,
      9,
      0,
      1,
      0,
      3,
      0,
      0,
      2,
      0,
      9,
      8,
      0,
      0,
      10,
      2,
      3,
      0,
    ],
  },
}
`;

exports[`4: open a document at \`./samples/fizzbuzz.nico\` 1`] = `
Object {
  "error": null,
  "id": 1001,
  "jsonrpc": "2.0",
  "result": Object {
    "data": Array [
      0,
      0,
      41,
      5,
      0,
      1,
      0,
      3,
      0,
      0,
      0,
      4,
      9,
      8,
      0,
      0,
      10,
      1,
      9,
      0,
      0,
      3,
      1,
      9,
      0,
      1,
      4,
      2,
      0,
      0,
      0,
      3,
      1,
      9,
      0,
      0,
      2,
      2,
      4,
      0,
      0,
      3,
      1,
      9,
      0,
      1,
      8,
      4,
      0,
      0,
      0,
      5,
      1,
      9,
      0,
      1,
      8,
      4,
      0,
      0,
      0,
      5,
      1,
      1,
      0,
      0,
      2,
      2,
      0,
      0,
      0,
      3,
      1,
      1,
      0,
      0,
      2,
      1,
      4,
      0,
      0,
      2,
      2,
      3,
      0,
      0,
      3,
      2,
      4,
      0,
      0,
      3,
      1,
      3,
      0,
      1,
      12,
      11,
      1,
      0,
      0,
      12,
      1,
      2,
      0,
      0,
      1,
      9,
      2,
      0,
      0,
      9,
      1,
      2,
      0,
      1,
      8,
      4,
      0,
      0,
      0,
      5,
      1,
      1,
      0,
      0,
      2,
      2,
      0,
      0,
      0,
      3,
      1,
      1,
      0,
      0,
      2,
      1,
      4,
      0,
      0,
      2,
      1,
      3,
      0,
      0,
      2,
      2,
      4,
      0,
      0,
      3,
      1,
      3,
      0,
      1,
      12,
      11,
      1,
      0,
      0,
      12,
      1,
      2,
      0,
      0,
      1,
      4,
      2,
      0,
      0,
      4,
      1,
      2,
      0,
      1,
      8,
      4,
      0,
      0,
      0,
      5,
      1,
      1,
      0,
      0,
      2,
      2,
      0,
      0,
      0,
      3,
      1,
      1,
      0,
      0,
      2,
      1,
      4,
      0,
      0,
      2,
      1,
      3,
      0,
      0,
      2,
      2,
      4,
      0,
      0,
      3,
      1,
      3,
      0,
      1,
      12,
      11,
      1,
      0,
      0,
      12,
      1,
      2,
      0,
      0,
      1,
      4,
      2,
      0,
      0,
      4,
      1,
      2,
      0,
      1,
      8,
      4,
      0,
      0,
      1,
      12,
      11,
      1,
      0,
      0,
      12,
      1,
      9,
      0,
      1,
      8,
      3,
      0,
      0,
      1,
      8,
      9,
      8,
      0,
      0,
      10,
      1,
      9,
      0,
      0,
      2,
      1,
      4,
      0,
      0,
      2,
      1,
      3,
      0,
      0,
      3,
      1,
      9,
      0,
      1,
      4,
      3,
      0,
      0,
      1,
      0,
      3,
      0,
      0,
      2,
      0,
      6,
      0,
      0,
      0,
      7,
      3,
      0,
      0,
      0,
      4,
      8,
      8,
      0,
      0,
      9,
      1,
      9,
      0,
      1,
      4,
      9,
      8,
      0,
      0,
      10,
      1,
      3,
      0,
      0,
      3,
      1,
      9,
      0,
      1,
      0,
      3,
      0,
      0,
      2,
      0,
      8,
      8,
      0,
      0,
      9,
      2,
      3,
      0,
    ],
  },
}
`;

exports[`5: open a document at \`struct Rectangle { width: i32, height: i32 }\` 1`] = `
Object {
  "error": null,
  "id": 1001,
  "jsonrpc": "2.0",
  "result": Object {
    "data": Array [
      0,
      0,
      6,
      0,
      0,
      0,
      7,
      9,
      1,
      0,
      0,
      12,
      5,
      1,
      0,
      0,
      12,
      6,
      1,
      0,
    ],
  },
}
`;
